### PR TITLE
#177 add increment on hold to number pickers

### LINKED ISSUE #177: Time Picker / Number Picker: Press and hold to increment/decrement
It should be easy enough to add press and hold detection to number pickers that let you hold down the button to increment/decrement.
This was raised as a feature request for time pickers when incrementing/decrementing the minutes.

UX Notes:
- There should be a slight delay between the mousedown and the start of the increment/decrement to account for long presses (around 200ms?)
- The increment speed should be reasonable enough that you can stop it on the intended number 90-100% of the time but not so slow that it's inconvenient (around 25ms?)
- It should be easier than clicking the button multiple times

### PR DESCRIPTION
**Pull request by c-frater**
resolves #177

### CODE REVIEW

**Comment by c-frater** [comment]
Need to add tests

File: modules/number-picker/main/index.coffee [hunk 1]
```diff
@@ -6,6 +6,24 @@ checkValue = (numberPicker, context) ->
...
 
+addHoldHandler = (incrementOnHold, incrementDelay, selection, incrementFn) ->
+  holdStart = undefined
+  holdTimer = undefined
+  incrementHold = undefined
+  if incrementOnHold
+    selection.on 'pointerdown', 'hx.number-picker', ->
+      holdStart = Date.now()
+      fn = -> incrementHold = setInterval((-> incrementFn()), incrementDelay)
+      holdTimer = setTimeout(fn, 200)
```
Comment (jonthans) [inline modules/number-picker/main/index.coffee]: Does this conflict with press-and-hold on touch devices?

**Comment by jonthans** [comment]
Can this get tested as well?

File: modules/number-picker/main/index.coffee [hunk 2]
```diff
@@ -6,6 +6,34 @@ checkValue = (numberPicker, context) ->
...
+    holdTimeout = undefined
+    incrementInterval = undefined
+
+    clearTimers = ->
+      clearTimeout(holdTimeout)
+      clearInterval(incrementInterval)
+
+    selection.on 'pointerdown', 'hx.number-picker', (e) ->
+      holdStart = Date.now()
+      document.activeElement.blur()
```
Comment (c-frater) [inline modules/number-picker/main/index.coffee]: Forcing blur here makes it possible to type in the input and then increment/decrement without having to click outside of the number picker first.

File: modules/number-picker/main/index.coffee [hunk 3]
```diff
@@ -6,6 +6,34 @@ checkValue = (numberPicker, context) ->
...
+      fn = ->
+        incrementInterval = setInterval((-> incrementFn()), incrementDelay)
+      holdTimeout = setTimeout(fn, 200)
+
+    selection.on 'pointerup', 'hx.number-picker', ->
+      clearTimers()
+      if (Date.now() - holdStart) < 200
+        incrementFn()
+
+    selection.on 'pointerleave', 'hx.number-picker', clearTimers
```
Comment (c-frater) [inline modules/number-picker/main/index.coffee]: This resolves an issue with desktops when moving the mouse out of selection and releasing outside.
This is a safer way of doing it than adding listeners to the body

File: modules/number-picker/main/index.coffee [hunk 4]
```diff
@@ -16,31 +44,30 @@ class NumberPicker extends hx.EventEmitter
...
-    @selectInput = select.append('input')
+    @selectInput = selection.append('input')
     @selectInput.attr('type', 'number')
     @selectInput.on 'blur', 'hx.number-picker', =>
-      if not @selectInput.attr('readonly')?
-        checkValue(this, @selectInput)
-        @selectInput.attr('data-value', @selectInput.value())
       @emit 'input-change', {value: @value()}
-      @emit 'change', {value: @value()}
+      @value(undefined, @selectInput.value())
```
Comment (c-frater) [inline modules/number-picker/main/index.coffee]: Value should be the only way to set the value

**Pull request merged**